EKF/terrain_estimator.cpp

```diff
@@ -134,40 +137,59 @@ void Ekf::fuseHagl()
 		float gate_size = fmaxf(_params.range_innov_gate, 1.0f);
 		_terr_test_ratio = sq(_hagl_innov) / (sq(gate_size) * _hagl_innov_var);
 
-		if (_terr_test_ratio <= 1.0f) {
-			// calculate the Kalman gain
-			float gain = _terrain_var / _hagl_innov_var;
-			// correct the state
-			_terrain_vpos -= gain * _hagl_innov;
-			// correct the variance
-			_terrain_var = fmaxf(_terrain_var * (1.0f - gain), 0.0f);
-			// record last successful fusion event
-			_time_last_hagl_fuse = _time_last_imu;
-			_innov_check_fail_status.flags.reject_hagl = false;
-
-		} else {
-			_innov_check_fail_status.flags.reject_hagl = true;
-
+		if (!_inhibit_gndobs_use) {
+			if (_terr_test_ratio <= 1.0f) {
+				// calculate the Kalman gain
+				float gain = _terrain_var / _hagl_innov_var;
+				// correct the state
+				_terrain_vpos -= gain * _hagl_innov;
+				// correct the variance
+				_terrain_var = fmaxf(_terrain_var * (1.0f - gain), 0.0f);
+				// record last successful fusion event
+				_time_last_hagl_fuse = _time_last_imu;
+				_innov_check_fail_status.flags.reject_hagl = false;
+
+			} else {
+				// If we have been rejecting range data for too long, reset to measurement
+				if (_time_last_imu - _time_last_hagl_fuse > (uint64_t)10E6) {
+					_terrain_vpos = _state.pos(2) + meas_hagl;
+					_terrain_var = obs_variance;
+				} else {
+					_innov_check_fail_status.flags.reject_hagl = true;
+				}
+			}
 		}
 
+
 	} else {
 		_innov_check_fail_status.flags.reject_hagl = true;
 		return;
 	}
 }
 
-// return true if the terrain estimate is valid
+// return true if the terrain height estimate is valid
 bool Ekf::get_terrain_valid()
 {
 	if (_terrain_initialised && _range_data_continuous && !_control_status.flags.rng_stuck &&
-		  (_time_last_imu - _time_last_hagl_fuse < (uint64_t)5e6)) {
+		  ((_time_last_imu - _time_last_hagl_fuse) < (uint64_t)5e6)) {
 		return true;
 
 	} else {
 		return false;
 	}
 }
 
+// determine terrain validity
+void Ekf::update_terrain_valid()
```

Comment:
This function does almost the same as `get_terrain_valid()`. Can we reuse that one to set `_hagl_valid` otherwise we need to add/remove checks in both locations if we change something.